doc/DIALER.md

```diff
@@ -0,0 +1,37 @@
+# js-libp2p Dialer
+
+**Synopsis**
+* All Dial Requests in js-libp2p must request a token(s) from the Dialer.
+  * The number of tokens requested should be between 1 and the PER_PEER_LIMIT max set in the Dialer.
+  * If the number of available tokens is less than requested, the Dialer may return less than requested.
+* The number of tokens a DialRequest obtains reflects the maximum number of parallel Multiaddr Dials it can make.
+* If no tokens are available a DialRequest should immediately end and throw. This deviates from the existing queue system to avoid queue congestion and provide more visibility to users.
+* As tokens are limited, DialRequests should be given a prioritized list of Multiaddrs to minimize the potential request time.
+* Once a single Multiaddr Dial has succeeded, all pending dials in that Dial Request should be aborted. All tokens should be immediately released to the Dialer.
+* If all Multiaddr Dials fail, or the DIAL_TIMEOUT max is reached for the entire DialRequest, all in progress dials for that DialRequest should be aborted. All tokens should immediately be released to the Dialer.
+* If a Multiaddr Dial fails and there are no more dials to use its token, that token should be immediately released to the Dialer.
+
+## Multiaddr Confidence
+
+**Not Yet Implemented**
+
+An effective dialing system should involve the inclusion of a Confidence system for Multiaddrs. This enables ranking of Multiaddrs so that a prioritized list can be passed to DialRequests to maximize usage of Dialer Tokens.
+
+* All Multiaddrs SHOULD start with a confidence of 1 (give it some trust by default)
```

Comment:
where are you planning to keep track of this? I know that we do not need to tackle it, but I would like to have this designed in a way that it will be easy to add this

Reply:
I would expect this to exist in the PeerStore when we separate PeerInfo into its components. That's where we should be tracking metadata like this, imo.

Reply:
I would expect this to exist in the PeerStore when we separate PeerInfo into its components. That's where we should be tracking metadata like this, imo.